src/scitacean/filesystem.py

```diff
@@ -0,0 +1,115 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2022 Scitacean contributors (https://github.com/SciCatProject/scitacean)
+# @author Jan-Lukas Wynen
+"""Filesystem utilities.
+
+Local paths are stored as pathlib.Path
+Remote paths are stored as scitacean.filesystem.RemotePath
+"""
+
+from __future__ import annotations
+
+import hashlib
+import os
+from datetime import datetime, timezone
+from pathlib import Path
+from typing import Callable, Generator, Optional, Union
+
+
+class RemotePath(os.PathLike):
+    """A path on the remote filesystem.
+
+    Remote paths need not correspond to a regular filesystem path like
+    :class:`pathlib.PosixPath` or :class:`pathlib.WindowsPath`.
+    Instead, they can be any sequence of segments that are joined by forward slashes,
+    e.g. a URL.
+    """
+
+    def __init__(self, path: Union[str, RemotePath]):
+        self._path = os.fspath(path)
+
+    def __truediv__(self, other: Union[str, RemotePath]) -> RemotePath:
+        """Join two path segments."""
+        this = _strip_trailing_slash(os.fspath(self))
+        other = _strip_leading_slash(_strip_trailing_slash(os.fspath(other)))
+        return RemotePath(f"{this}/{other}")
+
+    def __rtruediv__(self, other: str) -> RemotePath:
+        """Join two path segments."""
+        return RemotePath(other) / self
+
+    def __str__(self) -> str:
+        return self._path
+
+    def __repr__(self) -> str:
+        return f"RemotePath({str(self)})"
+
+    def __fspath__(self) -> str:
+        """Return the file system representation of the path."""
+        return str(self)
+
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, (RemotePath, str)):
+            return False
+        return str(self) == str(other)
+
+    def __hash__(self) -> int:
+        return hash(str(self))
```

Comment:
```suggestion
        return hash(self._path)
```
?

Reply:
Wanted to avoid relying on implementation details as much as possible just in case it changes to some other encoding. But I can change if yo uprefer.

Reply:
Well, but isn't relying on string formatting more risky? What if it eventually elides long paths, or something like that?